Labeling restructure

OriolAbril opened the pull request:
## Description
Enforcing and maintaining `index_origin` at label display has proved to be plenty of work and to not give the best of results. I propose to use `index_origin` at [InferenceData creation](https://github.com/arviz-devs/arviz/blob/master/arviz/data/base.py#L103) and then always display coordinate values. I think this will simplify the code significantly without many drawbacks. Moreover, this will make all representations of the data coherent, from xarray html repr to ArviZ plots.

Therefore, the aims of this PR are:
* [x] Use coordinate values in `az.summary`, fixes #1091 and fixes #1499
  * [x] Required restructuring `make_label` and `xarray_var_iter` to prevent cyclic imports

![image](https://user-images.githubusercontent.com/23738400/82715596-28185e80-9c94-11ea-9b39-9aa0a6b2cbcd.png)

* [x] Remove `index_origin` from stats and plotting functions. I basically affects 2 functions.
  * [x] `az.summary` 
  * [ ] `utils.flatten_inference_data_to_dict` and `to_cds` -> do in another PR, probably better to wait until we have sound converters for dict and dataframe? @percygautam @ahartikainen 
* [x] Use index_origin in default [coord generation](https://github.com/arviz-devs/arviz/blob/master/arviz/data/base.py#L103). Closes #984
* [x] Think (maybe implement) ways for users to customize generated labels options that seem interesting to have ~~are `var_name: coord_val`, `var_name [coord_val]`, `var_name [dim: coord_val]` and the like. Maybe a `label_str.format(var_name=var_name, dim=dim, coord_val=coord_val)` is enough?~~ Proposal below

## Proposal: Labeler class
I propose to use a labeler class with a handful of methods to cover the different labeling tasks. The base task is done with 2 functions, 1st apply the dim+coord labeling function looping over all present dims and 2nd combining the result of this function with the variable name. This could be a single `make_label` function, but I think a class will be better for extensibility, moreover, a class can also handle labeling when multiple models are present (i.e. plot_forest) and the different needs between plot labels and summary labels.

## Checklist
<!-- Feel free to remove check-list items aren't relevant to your change -->

- [x] Follows [official](https://github.com/arviz-devs/arviz/blob/master/CONTRIBUTING.md#pull-request-checklist) PR format
- [ ] New features are properly documented (with an example if appropriate)?
- [x] Includes new or updated tests to cover the new feature
- [x] Code style  correct (follows pylint and black guidelines)
- [x] Changes are listed in [changelog](https://github.com/arviz-devs/arviz/blob/master/CHANGELOG.md#v0xx-unreleased)

<!--
Also, please consider reading the contributing guidelines and code of conduct carefully before submitting the PR. They are available at
- https://github.com/arviz-devs/arviz/blob/master/CONTRIBUTING.md
- https://github.com/arviz-devs/arviz/blob/master/CODE_OF_CONDUCT.md

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in. Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

ahartikainen commented:
I'm not sure about this.

I kind of want to use 0-based indexing with pystan, and just have a summary with "correct" index.

Would this force me to use 1-based indexing?

OriolAbril commented:
Do you only want to see 1 based indexing in summary? or in plots too? 

Under the proposal above the coordinate values would be 1-base indexes, to use 0-based indexing you'd have to use `isel` instead of `sel`. Otherwise, to me it feels very weird to not be able to use the labels shown in summary to index which is what currently happens with index_origin=1.

ahartikainen commented:
Yeah, I forget isel. Yeah let's go forward with this

OriolAbril commented:
There are still some `from_xyz` functions to update, but I think the bulk of the PR is done. I think it is particularly relevant for users of `index_origin` which would mostly be @ahartikainen and @sethaxen. I think this will force ArviZ.jl to do a couple updates (I can help) and I do not expect too many things to break but it would be great to double check and more importantly to make sure this approach will help better integrate index_origin argument.

---

Tasks that will be left pending for other PRs are:
* update utils.flatten_inference_data_to_dict and to_cds to use coord labels (if possible)
* add function to customize label format
* testing?
  * Not sure if a couple tests on `generate_dims_coords` with index origin would be enough or if `from_xyz` should also be tested. any preference/idea @canyon289 ?

canyon289 commented:
@OriolAbril Can we create issue tickets for the remaining things and merge? Would hate for you to get stuck with a merge conflict as so many files have changed

OriolAbril commented:
I could be merged, but this introduces quite some changes in labeling and index_origin usage that affect most of the library. I think it would be better to wait until it has been properly reviewed. 

Most of the changes are reorganizing imports, so rebasing should be automatic or not too difficult, I don't mind a couple rebases until it gets some reviews

ahartikainen commented on arviz/data/base.py:
```diff
@@ -127,46 +139,62 @@ def numpy_to_data_array(ary, *, var_name="data", coords=None, dims=None):
 
-    if "chain" not in coords:
+    if "chain" not in coords and "chain" in default_dims:
         coords["chain"] = utils.arange(n_chains)
```
chains should still start from 0?

ahartikainen replied:
Or should it also start at index_origin

ahartikainen commented on arviz/sel_utils.py:
```diff
@@ -0,0 +1,224 @@
+        A text representation of the label
+    """
+    if selection:
+        sel = selection_to_string(selection)
```
just thinking about this

Let's say we have `theta[3]` with coords ["Ferrari", "Fiat", "Ford"]

Could we create labels as (can index origin work with these?)

    "theta[0] Ferrari"
    "theta[1] Fiat"
    "theta[2] Ford"

or 

    "theta[0]\nFerrari"
    "theta[1]\nFiat"
    "theta[2]\nFord"

For this to happen, do we need to update selection (`xarray_sel_iter`?). And will index_origin work when coordinates are explicitly set?

OriolAbril replied:
I want this to be possible and should be if we support custom labeling function, I am not sure what would be a good way though. 

I do think we will have to update xarray_sel_iter or use xarray_sel_iter to create a function targeted to labeling. In addition to the index, I also would like to expose the dimension so that `theta\nbrand: Ferrari`, `theta\nbrand: Fiat` are also possible. It may also be interesting to keep coordinate values simple (maybe abbreviations) and then have some attribute `label` (and `labels` attr in coordinates) so that users could select variable `theta` using `theta` but set its name to `$\theta$` for plots to show the greek letter. This gets crazy quite fast though. Eventually, `var_name`, `dim`/`coord_name`, `coord_value`, `coord_idx` and maybe `var_label`, `dim_label` and `coord_label` would be exposed when creating labels so custom functions can play with them to get the desired labels.

Regarding `coord_idx`, I think it should not include `index_origin` to get the same value that would be used by `isel`, but we should have an example on adding `index_origin` to `coord_idx` inside the labeller function.

ahartikainen commented on arviz/sel_utils.py:
```diff
@@ -0,0 +1,224 @@
+    if selection:
+        sel = selection_to_string(selection)
+        if position == "below":
+            base = "{}\n{}"
```
This looks bad with numbers, but good with strings

OriolAbril replied:
yep, any ideas on how to handle this are very welcome, having these 2 different defaults (one for plots and another for summary) also makes it harder (at least for me) to think about sensible ways to change the labeling function

ahartikainen commented on arviz/sel_utils.py:
```diff
@@ -0,0 +1,224 @@
+        if position == "below":
+            base = "{}\n{}"
+        elif position == "beside":
+            base = "{}[{}]"
```
this is correct with numbers, "bad" with strings

ahartikainen commented on arviz/sel_utils.py:
this should somehow return also the indices (and respect the index origin)

OriolAbril replied:
I agree on returning indices, I would not enforce index_origin here (the coordinate label would respect index_origin if integer). I personally don't like having indices that do not correspond to the ones used to index/slice and therefore, as I said above, I would use regular indices and eventually (and optionally) add `index_origin`

ahartikainen replied:
That sounds more robust way to handle this

ahartikainen commented on arviz/sel_utils.py:
selection should be tuple with `iloc` and `loc` selection methods, right?

OriolAbril replied:
that sounds good, how about `yield var_name, selection, iselection, data_to_sel...`?

ahartikainen replied:
that option sounds good

OriolAbril replied:
I was looking into this and I realized that having indexes is currently close to impossible, I think it would be better to move this to an issue and also discuss what we want to be possible in a lab meeting.

One of the first operations done is `dataset.sel()`, so the correct index information is lost before calling `xarray_sel_iter`.

ahartikainen commented on arviz/stats/stats.py:
```diff
@@ -1045,6 +1040,11 @@ def summary(
     credible_interval: float, optional
         deprecated: Please see hdi_prob
+    order
+        deprecated: order is now ignored.
```
Why is order removed?

ahartikainen replied:
or is there a way to select order for unwrapped values?

OriolAbril replied:
I think that having informative labels, order is not really necessary (also it is not present in any plot and nobody has complained). High dimension and confusing cases should probably better use `dim: coord_value` format or something like this. 

One of the goals here is to make summary and plots use the same labeling and between removing order from summary or adding order to plots I went for removing order from summary.

rpgoldman replied:
Weren't we talking about returning xarray's of axes (or other backend entities) as a way to do this?  TBH, I don't know if it's possible to move around the subplots in matplotlib -- if this can be done, then the xarray solution would work.

canyon289 commented:
Hi, are there things we can do to help get this one merged? Looks like a lot of work

OriolAbril commented:
I think the current code is mostly ok and is an improvement over the actual one, not sure it's the way to go though, it still has some clear limitations and it's hard to customize the label formatting.

ahartikainen commented:
this can wait, right

OriolAbril commented:
I have rebased the PR and continued the work started here. I proposed to use a Labeller class. The latest commit adds the base class and some subcasses I think will be interesting. I have added some screenshots below, please review.

## BaseLabeller (used by default)

![image](https://user-images.githubusercontent.com/23738400/108285309-24fcc100-7187-11eb-83d7-a078d4a34c76.png)

## DimCoordLabeller

![image](https://user-images.githubusercontent.com/23738400/108285365-3fcf3580-7187-11eb-82b5-32c2367f8732.png)

## NoRepeatLabeller

![image](https://user-images.githubusercontent.com/23738400/108285433-5d9c9a80-7187-11eb-8ffa-6866401660b7.png)

## MapLabeller

![image](https://user-images.githubusercontent.com/23738400/108285481-74db8800-7187-11eb-8ba8-94926a31d533.png)

OriolAbril commented on arviz/labels.py:
```diff
@@ -0,0 +1,70 @@
+
+    def dim_coord_to_str(self, dim, coord_val, coord_idx):
+        dim_str = self.dim_map.get(dim, dim)
+        coord_str = self.coord_map.get(coord_val, coord_val)
```
This could also be `coord_str = self.coord_map.get(dim, {}).get(coord_val, coord_val)`, it won't be as conveninet when there is only one dimension present, but for multidim it may be the only way

OriolAbril commented on arviz/sel_utils.py:
```suggestion
            dims = _zip_dims(new_dims, vals)
            idims = _zip_dims(new_dims, [np.arange(len(v)) for v in vals])
            if reverse_selections:
                dims = reversed(dims)
                idims = reversed(idims)

            for selection in dims:
                yield var_name, selection, iselection
```

@ahartikainen I think this will correctly return the index, not sure if it will reduce performance noticeably though. And it has an important caveat which can't be overcome for now (but I think labeller class can fix it) as it's a construction issue as I commented at some point before. One of the first steps is to call `get_coords` to subset the data if `coords` arg is present. And we would take the positional indexes from there, not from the original dataset. 

long story short, when coords are passed, the shown positional indexes would work for `dataset.sel(coords).isel(...)` but will not necessarily work with `dataset.isel(...)`. I think its fine if we document it, and there may be ways to overcome that with labeller class.

ahartikainen replied:
I think you need

    for selection, iselection in zip(dims, idims)

OriolAbril commented on arviz/plots/bpvplot.py:
```diff
@@ -93,6 +96,8 @@ def plot_bpv(
         For example, `data_pairs = {'y' : 'y_hat'}`
         If None, it will assume that the observed data and the posterior/prior
         predictive data have the same variable name.
+    labeller : labeller instance, optional
+        Class providing the method `make_label_vert` to generate the labels in the plot titles.
```
Do we want to allow only labeller objects? Or strings too and have some kind of mapping/`getattr(az.labels, string)`?

OriolAbril commented on arviz/plots/backends/bokeh/bpvplot.py:
```diff
@@ -175,12 +176,12 @@ def plot_bpv(
-        else:
-            xlabel = var_name
+        # if var_name != pp_var_name:
+        #     xlabel = "{} / {}".format(var_name, pp_var_name)
+        # else:
+        #     xlabel = var_name
         _title = Title()
-        _title.text = xlabel
+        _title.text = labeller.make_label_vert(var_name, sel, isel)
```
Here matplotlib and bokeh behaved diferently. I have modified bokeh to match mpl, but we can also add a new method to the labellers to take the possible 2 var_names into account. cc @aloctavodia

OriolAbril commented:
Updated several plots to see how labeling behaved in different ranges, and while I was at it I added a legend argument to `plot_forest`. I have set it to true by default, which combined with #1527 leaves this as default:

![image](https://user-images.githubusercontent.com/23738400/108440533-9bff8b80-7253-11eb-8463-993d9bc1eb4f.png)

The rest of the plots work more or less like summary above but with plots, note that the screenshots have the wrong default though, we are now using `var_name[coord]`,  `var_name[dim: coord]` or `var_name[dim#idx]`.

sethaxen commented:
> I think it is particularly relevant for users of `index_origin` which would mostly be @ahartikainen and @sethaxen. I think this will force ArviZ.jl to do a couple updates (I can help) and I do not expect too many things to break but it would be great to double check and more importantly to make sure this approach will help better integrate index_origin argument.

What kinds of things do you think might need to be updated on the ArviZ.jl side?

OriolAbril commented:
> What kinds of things do you think might need to be updated on the ArviZ.jl side?

There are two changes that I think will affect the Julia side, on the user side it should be backwards compatible in the sense that the noticeable changes will be fixing issues.

1. index_origin things. Up until this PR, when no coord values were provided, the converters generated automatic integer values starting from 0. Then, plots and stats had an `index_origin` parameter that added 1 to interger coord values so that even though the coord values started at 0, plot labels would start at 1. 

From here on, users can choose what to use when labeling, coords, indexes, nothing at all... and the default is set to always use coord values. index_origin is then enforced at idata creation. Therefore, stan users will set index_origin=1 and will get coordinate values starting from 1. They can then subset their data, chains, draws... with `.sel` using 1-based indexing or with `.isel` using 0-based indexing.

To keep showing 1 based indexing in plots, the julia converters should be updated to enforce index_origin at creation.

2. is about index based labeling. Up until now, the info was simply not available to plotting functions, now it is, which means that  `xarray_var_iter` and `xarray_sel_iter` return 4 elements now instead of 3. 

I don't know if these are used at all on the julia side, but if they are everything will break :/

If you point me to the relevant pieces of code I can help with updating

sethaxen commented:
> 1. index_origin things. Up until this PR, when no coord values were provided, the converters generated automatic integer values starting from 0. Then, plots and stats had an `index_origin` parameter that added 1 to interger coord values so that even though the coord values started at 0, plot labels would start at 1.
> 
> From here on, users can choose what to use when labeling, coords, indexes, nothing at all... and the default is set to always use coord values. index_origin is then enforced at idata creation. Therefore, stan users will set index_origin=1 and will get coordinate values starting from 1. They can then subset their data, chains, draws... with `.sel` using 1-based indexing or with `.isel` using 0-based indexing.
> 
> To keep showing 1 based indexing in plots, the julia converters should be updated to enforce index_origin at creation.

Nice! Currently ArviZ.jl handles this by setting the `data.index_origin` rcParam upon loading arviz: 
https://github.com/arviz-devs/ArviZ.jl/blob/v0.4.11/src/setup.jl#L55
If the user doesn't supply `index_origin` upon creation, does it default to the rcParam, or is the rcParam going away?

> 1. is about index based labeling. Up until now, the info was simply not available to plotting functions, now it is, which means that  `xarray_var_iter` and `xarray_sel_iter` return 4 elements now instead of 3.
> 
> I don't know if these are used at all on the julia side, but if they are everything will break :/

No, we don't use these on the Julia side, though ArviZPlots currently does. Should be easy to update though; I'll follow how you do that for arviz's plots implementations.

OriolAbril commented:
> If the user doesn't supply index_origin upon creation, does it default to the rcParam, or is the rcParam going away?

The rcParam is kept so users can set the default there and get the right coord values without having to set index_origin every single time they call the converter.

sethaxen commented:
> The rcParam is kept so users can set the default there and get the right coord values without having to set index_origin every single time they call the converter.

Okay, then it sounds like for ArviZ.jl no changes will be necessary. I'll add to our converters the new `index_origin` argument and update ArviZPlots.jl accordingly when a new release includes these changes.

OriolAbril commented:
Ready to merge as soon as tests pass!
